### PR TITLE
feat(changelog): Add support for sub-sections

### PR DESCRIPTION
**Pull request by pree**
Relates to https://github.com/CramBL/dependabot-changelog-writer/issues/114

As mentioned, I'm a rust newbie and trying to implement this. It's not fully functional yet, so I'm opening it as a draft PR. Feel free to already poke at thinks in case I'm off to a completely wrong direction.

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -72,7 +76,7 @@ runs:
       shell: bash
       run: |
         echo "action_repo=dependabot-changelog-writer" >> $GITHUB_ENV
-        echo "action_org=CramBL" >> $GITHUB_ENV
+        echo "action_org=pree" >> $GITHUB_ENV
```
Comment (CramBL) [inline action.yml:79]:
```suggestion
        echo "action_org=CramBL" >> $GITHUB_ENV
```

File: src/changelog/parse.rs [hunk 2]
```diff
@@ -190,6 +190,48 @@ pub fn find_new_h3_insert_position(changelog_content: &str) -> usize {
...
+    let mut content_pos = 0;
+    for l in changelog_content.split_inclusive('\n') {
+        content_pos += l.len();
+        if l.starts_with("-") {
+            if l[1..].contains(sub_section_header){
+                let mut offset_withing_sub_section = 0;
+                for l in changelog_content[content_pos..].split_inclusive('\n') {
+                    if l.starts_with("-") {
+                        // There is usually no empty line between list items
+                        // offset_withing_subsection -= 1;
```
Comment (CramBL) [inline src/changelog/parse.rs]:
```suggestion
```

File: src/changelog/parse.rs [hunk 3]
```diff
@@ -190,6 +190,48 @@ pub fn find_new_h3_insert_position(changelog_content: &str) -> usize {
...
+//    for l in changelog_content.split_inclusive('\n') {
+//        // Check for h3 header
+//        if l.starts_with("###") {
+//            return content_pos;
+//        }
+//        content_pos += l.len();
+//    }
+//    content_pos
+//}
+
```
Comment (CramBL) [inline src/changelog/parse.rs]:
```suggestion
```

File: src/dependabot_changes/entry_pattern.rs [hunk 4]
```diff
@@ -91,7 +91,8 @@ impl EntryPattern {
                 Self::PULL_REQUEST_LINK_TOKEN_HARDENED,
             );
 
-        let mut cooked_pattern = Self::LINE_PREFIX.to_owned();
+        let mut cooked_pattern = if indentation { "  ".to_owned() } else { "".to_owned() };
```
Comment (CramBL) [inline src/dependabot_changes/entry_pattern.rs]: Indentation is a pretty controversial topic, how did you arrive at 2 spaces?

Unless there's a very good reason for 2 spaces, I think we have to support any number of spaces (some like 2 spaces, some like 3 or 4, we might as well support any number) and tabs.

So I would have the action.yml have a feature such as "sub-section-header-indent" and have it take values such as "NUM-spaces" e.g. "2-spaces" or "tab". And just default to 2-spaces I guess?

**Review by CramBL — CHANGES_REQUESTED** [review]
Looks good, you're like 95% of the way there.

I have some comments, and then there's some typos to fix, and finally this needs some tests as well.

At least a test for using the feature to place content in a sub section, and also a test where there's an existing subsection, to ensure that it correctly navigates to that sub section and places contents in the expected position.

File: action.yml (hunk 1 above)
Comment (pree) [reply action.yml:79]: Yes, this is just for my testing, will change this back before marking it as ready again :)

File: src/dependabot_changes/entry_pattern.rs (hunk 4 above)
Comment (pree) [reply src/dependabot_changes/entry_pattern.rs]: Good idea, I think everyone has different opinions about this. [GitHub's Markdown Syntax](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#nested-lists) doesn't specify this either.

**Comment by pree** [comment]
@CramBL Thanks for the initial review already! I'll be working on this again next week.

**Comment by pree** [comment]
I think I've got it into a working stage, took me way longer than I expected. I guess learning a new language is not as easy as one would envision :D 

I will be adding some more tests, as I'm sure there is some edge cases were it would fail.
